File: lambdaconscarcdr.pl
#!/usr/bin/env perl

use strict;
use warnings;
use feature 'signatures';

my $kons  = sub ($hd, $tl = undef) {
  sub ($x) { if ($x) {$hd} else {$tl} }
};

my $virst = sub { shift->(1    ) };
my $rrest = sub { shift->(undef) };

my $lstt  = $kons->(1, $kons->(2));

$virst->($rrest->($lstt))

File: lambdaconscarcdr.m
#!/usr/bin/env octave

function retval = kons (hd, tl)
  retval = @ (x) ifelse (x, hd, tl)
endfunction

function retval = virst (l)
  retval = l(true )
endfunction

function retval = rrest (l)
  retval = l(false)
endfunction

lstt = kons (1, kons (2, NA))

virst (rrest (lstt))
